p2p/transport.go

```diff
@@ -0,0 +1,579 @@
+	return func(mt *MultiplexTransport) { mt.connFilters = filters }
+}
+
+// MultiplexTransportFilterTimeout sets the timeout waited for filter calls to
+// return.
+func MultiplexTransportFilterTimeout(
+	timeout time.Duration,
+) MultiplexTransportOption {
+	return func(mt *MultiplexTransport) { mt.filterTimeout = timeout }
+}
+
+// MultiplexTransportPeerFilters sets the filters for rejection of new peers.
+func MultiplexTransportPeerFilters(
+	filters ...PeerFilterFunc,
+) MultiplexTransportOption {
+	return func(mt *MultiplexTransport) { mt.peerFilters = filters }
+}
+
+// MultiplexTransport accepts and dials tcp connections and upgrades them to
+// multiplexed peers.
+type MultiplexTransport struct {
+	listener net.Listener
+
+	acceptc chan accept
+	closec  chan struct{}
+
+	// Lookup table for duplicate ip and id checks.
+	conns    map[string]net.Conn
+	connsMux sync.RWMutex
+	peers    map[ID]Peer
+	peersMux sync.RWMutex
+
+	dialTimeout      time.Duration
+	filterTimeout    time.Duration
+	handshakeTimeout time.Duration
+	nodeAddr         NetAddress
+	nodeInfo         NodeInfo
+	nodeKey          NodeKey
+
+	connFilters []ConnFilterFunc
+	peerFilters []PeerFilterFunc
+
+	// TODO(xla): Those configs are still needed as we parameterise peerConn and
+	// peer currently. All relevant configuration should be refactored into options
+	// with sane defaults.
+	mConfig   conn.MConnConfig
+	p2pConfig config.P2PConfig
+}
+
+// Test multiplexTransport for interface completeness.
+var _ Transport = (*MultiplexTransport)(nil)
+var _ transportLifecycle = (*MultiplexTransport)(nil)
+
+// NewMultiplexTransport returns a tcp conencted multiplexed peers.
+func NewMultiplexTransport(
+	nodeInfo NodeInfo,
+	nodeKey NodeKey,
+) *MultiplexTransport {
+	return &MultiplexTransport{
+		acceptc:          make(chan accept),
+		closec:           make(chan struct{}),
+		dialTimeout:      defaultDialTimeout,
+		filterTimeout:    defaultFilterTimeout,
+		handshakeTimeout: defaultHandshakeTimeout,
+		mConfig:          conn.DefaultMConnConfig(),
+		nodeInfo:         nodeInfo,
+		nodeKey:          nodeKey,
+		conns:            map[string]net.Conn{},
+		peers:            map[ID]Peer{},
+	}
+}
+
+// Accept implements Transport.
+func (mt *MultiplexTransport) Accept(cfg peerConfig) (Peer, error) {
+	select {
+	// This case should never have any side-effectful/blocking operations to
+	// ensure that quality peers are ready to be used.
+	case a := <-mt.acceptc:
+		if a.err != nil {
+			return nil, a.err
+		}
+
+		cfg.outbound = false
+
+		p := mt.wrapPeer(a.conn, a.nodeInfo, cfg)
+
+		return p, mt.filterPeer(p)
+	case <-mt.closec:
+		return nil, &ErrTransportClosed{}
+	}
+}
+
+// Dial implements Transport.
+func (mt *MultiplexTransport) Dial(
+	addr NetAddress,
+	cfg peerConfig,
+) (Peer, error) {
+	c, err := addr.DialTimeout(mt.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(xla): Evaluate if we should apply filters if we explicitly dial.
```

Comment:
Good question. We probably want to prevent ourselves from dialing duplicates as much as possible in the first place, but at least for now it might happen if mechanisms like PEX and persistentPeers are concurrently trying to dial the same peers